Sources/SwiftKafka/KafkaProducer.swift

```diff
@@ -14,21 +14,289 @@
         logger: Logger
-    ) throws {
-        self.client = try KafkaClient(type: .producer, config: config, logger: logger)
+    ) async throws {
+        self.config = config
         self.topicConfig = topicConfig
+        self.logger = logger
+        self.topicHandles = [:]
+        self.state = .started
+
+        // (NIOAsyncSequenceProducer.makeSequence Documentation Excerpt)
+        // This method returns a struct containing a NIOAsyncSequenceProducer.Source and a NIOAsyncSequenceProducer.
+        // The source MUST be held by the caller and used to signal new elements or finish.
+        // The sequence MUST be passed to the actual consumer and MUST NOT be held by the caller.
+        // This is due to the fact that deiniting the sequence is used as part of a trigger to
+        // terminate the underlying source.
+        let acknowledgementsSourceAndSequence = NIOAsyncSequenceProducer.makeSequence(
+            of: Acknowledgement.self,
+            backPressureStrategy: NoBackPressure(),
+            delegate: NoDelegate()
+        )
+        self.acknowledgements = AcknowledgedMessagesAsyncSequence(
+            wrappedSequence: acknowledgementsSourceAndSequence.sequence
+        )
+
+        self.callbackOpaque = OpaqueWrapper(
+            source: acknowledgementsSourceAndSequence.source,
+            logger: self.logger
+        )
+
+        self.config.setDeliveryReportCallback(callback: self.deliveryReportCallback)
+        self.config.setOpaque(opaque: self.callbackOpaque)
+
+        self.client = try KafkaClient(type: .producer, config: self.config, logger: self.logger)
+
+        // Poll Kafka every millisecond
+        self.pollTask = Task { [client] in
+            while !Task.isCancelled {
+                client.withKafkaHandlePointer { handle in
+                    rd_kafka_poll(handle, 0)
+                }
+                try? await Task.sleep(nanoseconds: 1_000_000)
+            }
+        }
+    }
+
+    /// Method to shutdown the ``KafkaProducer``.
+    ///
+    /// This method flushes any buffered messages and waits until a callback is received for all of them.
+    /// Afterwards, it shuts down the connection to Kafka and cleans any remaining state up.
+    public func shutdownGracefully() throws {
+        switch self.state {
+        case .started:
+            self.state = .shuttingDown
+            self._shutDownGracefully()
+        case .shuttingDown, .shutDown:
+            return
+        }
+    }
+
+    // TODO: put to deinit, remove state?
+    private func _shutDownGracefully() {
+        Task {
+            // Wait 10 seconds for outstanding messages to be sent and callbacks to be called
+            self.client.withKafkaHandlePointer { handle in
+                rd_kafka_flush(handle, 10000)
+            }
+        }
+
+        for (_, topicHandle) in self.topicHandles {
+            rd_kafka_topic_destroy(topicHandle)
+        }
+        self.pollTask.cancel()
```

Comment:
This doesn't seem right to me. We are flushing on a `Task` and concurrently destroying all the topic handles. Is that correct?

Reply:
Uhhh yeah, *dumb* mistake — what do you think of getting rid of the `Task` and wrapping this blocking call with `await`? I guess that is what @Lukasa meant. Furthermore, we should think about if we even need a `shutDownGracefully()` method when we do not keep a reference to `self` in our `KafkaProducer`. We could then just make use of `deinit`. What do you think?

Reply:
We could potentially leave out the `flush` function call. However, this could lead to some queued messages getting lost without the user noticing. As far as I understand, the Kafka C library manages a message queue for outgoing messages. We can only put messages into this queue without influencing when they are sent. Therefore this is a safety mechanism to avoid messages in the queue getting lost when we close the producer.

Reply:
It might make sense to let the user define that deadline. Instead of us fixing it to 10s

Reply:
Yes, let's do this! We can still stick with 10s as a default value

Reply:
Have force-pushed a `timeout` parameter for `shutDownGracefully`, just to let you know